samza-rest/src/main/java/org/apache/samza/rest/proxy/job/YarnRestJobStatusProvider.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.rest.proxy.job;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.codec.binary.StringUtils;
+import org.apache.commons.httpclient.HttpClient;
+import org.apache.commons.httpclient.HttpStatus;
+import org.apache.commons.httpclient.methods.GetMethod;
+import org.apache.hadoop.yarn.api.records.YarnApplicationState;
+import org.apache.samza.SamzaException;
+import org.apache.samza.rest.model.yarn.ApplicationInfo;
+import org.apache.samza.rest.model.Job;
+import org.apache.samza.rest.model.JobStatus;
+import org.apache.samza.rest.resources.JobsResourceConfig;
+import org.apache.samza.rest.resources.YarnJobResourceConfig;
+import org.codehaus.jackson.map.DeserializationConfig;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of the {@link JobStatusProvider} that retrieves
+ * the job status from the YARN REST api.
+ */
+public class YarnRestJobStatusProvider implements JobStatusProvider {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(YarnRestJobStatusProvider.class);
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static final String JOB_NAME_ID_FORMAT = "%s_%s";
+  private final String apiEndpoint;
+  private final HttpClient httpClient;
+
+  public YarnRestJobStatusProvider(JobsResourceConfig config) {
+    YarnJobResourceConfig yarnConfig = new YarnJobResourceConfig(config);
+    this.httpClient = new HttpClient();
```

Comment:
Why not use YarnClient?
https://github.com/apache/hadoop/blob/trunk/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/main/java/org/apache/hadoop/yarn/client/api/YarnClient.java

Then we wouldn't need the ApplicationInfo classes at all. We could leverage Yarn's ApplicationReport.

Reply:
Built a working prototype of the same provider using YarnClient. It seems to work perfectly fine as long as the RM address points to an Active RM.
If the RM endpoint is configured to a Passive RM, YarnClient throws a `java.net.ConnectException: Connection refused`.

On the other hand, hitting the YARN Http REST endpoint on a Passive RM returns a HTTP 302 redirect. Therefore, it always works regardless of which RM's (Active/Passive) JobsResource endpoint is hit, but this would require us to maintain models similar to what YarnClient already gives us by default.

Not sure how to proceed here.

Reply:
Discussed in person. There doesn't seem to be any simple workaround for the YarnClient approach, so I'm ok with the HttpClient. It's still better than the CLI implementation.